### PR TITLE
Article: How to JSON with Flutter

### PR DESCRIPTION
**Pull request by roughike**
Fixes flutter/flutter#13520.

### CODE REVIEW

File: json.md [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+  # Your other dev_dependencies here
+  build_runner: ^0.6.1
+  json_serializable: ^0.3.0
+```
+
+After declaring the needed dependencies, synchronize your packages. In IntelliJ IDEA, click the _Packages get_ link on the top of an opened `pubspec.yaml` file. Alternatively, you can do synchronize packages by running `flutter packages get` from the command line in your projects' root folder.
+
+Next,  define some build actions.
+
+#### Defining the build actions
```
Comment (kevmoo) [inline json.md]: FYI – this has been simplified a lot.

See https://github.com/dart-lang/json_serializable/tree/master/example

Comment (roughike) [reply json.md]: Great! I'll look into it and update this section.

Comment (kevmoo) [reply json.md]: Awesome.

Comment (roughike) [reply json.md]: I tried to take a quick look into it, and ran into the exact same issue that [this reddit user describes in this comment](https://www.reddit.com/r/FlutterDev/comments/7nymf5/simple_json_with_flutter/ds80p88/).

This is reproducible on the latest Flutter version, and happens by simply following these steps:

* create a new Flutter project
* add the dependencies listed on the `json_serializable` README
* attempt to get the packages -> a long error message which states that a version `^0.7.0` for `build_runner` was found.

Should I create a more detailed issue report, or is this some common issue that can be resolved easily?

File: _includes/sidebar.html [hunk 2]
```diff
@@ -38,6 +38,7 @@
       <li><a href="/platform-channels/">Platform-specific code</a></li>
       <li><a href="/reading-writing-files/">Read and write files</a></li>
       <li><a href="/networking/">Network and HTTP</a></li>
+      <li><a href="/json">Serializing JSON objects</a></li>
```
Comment (sethladd) [inline _includes/sidebar.html]: maybe more simply: "JSON and serialization"

File: json.md [hunk 3]
```diff
@@ -0,0 +1,301 @@
+---
+layout: page
+title: TODO - JSON article title
+permalink: /json/
+---
+
+It is hard to think of a mobile app that doesn't need to communicate with a web server at some point. When making network-connected apps, chances are that we need to consume some good old JSON, sooner or later.
```
Comment (sethladd) [inline json.md]: global: please line break at 80 chars

File: json.md [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+
+In this tutorial, we look into ways of doing JSON with Flutter. We will go over what JSON solution to use in different scenarios and why.
```
Comment (sethladd) [inline json.md]: s/doing/using

File: json.md [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+Different projects come with different complexities and use cases. For smaller proof-of-concept projects or quick prototypes, using code generators might be overkill. For apps with several JSON models with more complexity, doing serialization by hand can quickly become tedious, repetitive and lend itself to lots of small errors.
+
+### Use manual serialization for smaller projects
+
+Manual JSON serialization refers to using the built-in JSON decoder in `dart:convert`. It involves passing the raw JSON string to the `JSON.decode()` method, and then looking up the values you need in the `Map<String, dynamic>` the method returns. It involves no external dependencies or particular setup process, and it is good for quick proof of concepts.
+
+Where the manual serialization does not perform well is when your project becomes bigger. Writing the serialization logic by hand can become hard to manage and error-prone. If you have a typo when accessing an unexisting JSON field, your code throws an error during runtime.
+
+If you do not have many JSON models in your project and are looking to test a concept quickly, manual serialization might be the way you want to start.
```
Comment (sethladd) [inline json.md]: link down to the example section. For example: "To learn more, and to see an example, jump to ___________"

(global comment, link to each example from these intro paragraphs)

File: json.md [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+### Use code generation for medium to large projects 
+
+JSON serialization with code generation means having an external library generate the serialization boilerplate for you. They involve some initial setup and running a file watcher that generates the code from your model classes. For example, [json_serializable](https://pub.dartlang.org/packages/json_serializable) and [built_value](https://pub.dartlang.org/packages/built_value) are these kind of libraries.
+
+This approach scales well for a larger project. There are no hand-written boilerplate needed, and typos when accessing JSON fields are caught at compile-time. The downside with code generation is that it involves some initial setup. Also, the generated source files may produce visual clutter in your project navigator
+
+You might want to use generated code for JSON serialization when you have a medium or a larger project.
+
+## Is there a GSON/Jackson/Moshi equivalent in Flutter?
+
```
Comment (sethladd) [inline json.md]: Answer the question immediately, and then provide the background :)

Comment (sethladd) [reply json.md]: it's hard to tell: yes, or no

File: json.md [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+Dart has supported _tree shaking_ for quite a long time. With tree shaking, we can “shake off” unused code from our release builds. Tree shaking allows us to optimize the size of our applications significantly.
+
+While Dart users might have gotten used to [dartson](https://pub.dartlang.org/packages/dartson) to (de)serialize their JSON, it does not work with Flutter. The reason behind this is that `dartson`, like [GSON](https://github.com/google/gson) in Java, uses _runtime reflection_. In Flutter, reflection (also known as `dart:mirrors`) is disabled.
```
Comment (sethladd) [inline json.md]: I might make this paragraph more of an aside. It's probably not relevant to most users. Can you make it an aside, and get to one sentence? e.g. "What about dartson? ____________"

File: json.md [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+
+Since reflection makes all code implicitly used by default, it interferes with tree shaking. The tools cannot know what parts are unused at runtime; the redundant code is impossible to strip away. App sizes cannot be optimized when using reflection.
+
+Although we cannot use dartson, there’s a library that gives us a simple to use API, without using reflection. That library is called [json_serializable](https://pub.dartlang.org/packages/json_serializable).
+
+## Serializing JSON manually using dart:convert
+
+Basic JSON serialization in Flutter is very simple. Dart comes with the `dart:convert` library built in, and using the JSON decoder API is straightforward.
```
Comment (sethladd) [inline json.md]: "comes with" and "built in" are redundant

Comment (sethladd) [reply json.md]: maybe: "Flutter has a built-in 'dart:convert' library, which includes a straightforward JSON encoder and decoder."

File: json.md [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+
+Let's say we need to deserialize a JSON representing a user object. For sample purposes, we have a real simple JSON model.
```
Comment (sethladd) [inline json.md]: "Here is an example JSON for a simple user." to replace the second paragraph

File: json.md [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+**user.dart**
+
+```dart
+class User {
+  User(this.name, this.email);
+
+  String name;
+  String email;
+
+  User.fromJson(Map<String, dynamic> json)
```
Comment (sethladd) [inline json.md]: should we not put the constructors together?

File: json.md [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+
+This way, the calling code can now have type safety, autocompletion for the `name` and `email` fields and compile-time exceptions. If we make typos or treat the fields as `int`s instead of `String`s, instead of crashing on runtime, our app will not even compile.
+
+**user.dart**
+
+```dart
+class User {
+  User(this.name, this.email);
+
+  String name;
```
Comment (sethladd) [inline json.md]: let's make these final

File: json.md [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+    };
+  }
+}
+```
+
+Now the calling code does not have to care about the serialization logic. With this new approach, we can deserialize a user quite easily.
+
+```dart
+Map map = JSON.decode(json);
+User user = new User.fromJson(map);
```
Comment (sethladd) [inline json.md]: s/User user/var user (since it's obvious

File: json.md [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+      'email': email,
+    };
+  }
+}
+```
+
+Now the calling code does not have to care about the serialization logic. With this new approach, we can deserialize a user quite easily.
+
+```dart
+Map map = JSON.decode(json);
```
Comment (sethladd) [inline json.md]: is there a better variable name for map ?

File: json.md [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+print('We sent the verification link to ${user.email}.');
+```
+
+To serialize a user, we just call the `toJson` method in the `User` class.
+
+```dart
+String json = JSON.encode(user.toJson());
+```
+
+While with this new approach the calling code does not have to worry about JSON serialization at all, the model class still definitely has. In a production app, we would want to be sure that the serialization works. In practice, this means that the `User.fromJson` and `User.toJson` both need to have unit tests in place to verify correct behavior.
```
Comment (sethladd) [inline json.md]: this first sentence was hard to parse.

File: json.md [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+
+Also, real-world scenarios are not usually that simple. It is unlikely that we can get by with such small JSON responses. Nested JSON objects are not that uncommon either.
+
+It would be nice if there were something that handled the JSON serialization for us. Luckily, there is!
+
+## Serializing JSON using json_serializable
+
+The `json_serializable` package is an automated source code generator that can generate the JSON serializing boilerplate for us. 
+
+Since the serialization code is not handwritten and maintained by us anymore, we minimize the risk of having JSON serialization exceptions at runtime. If the sources are generated, and the app compiles, we should be fine.
```
Comment (sethladd) [inline json.md]: I think this second sentence is redundant.

**Comment by sethladd** [comment]
Fantastic! Thank you so much.

File: json.md (hunk 1 above)
Comment (kevmoo) [reply json.md]: We're in dependency lock, I'm pretty sure.

I think I have a path to untangle the knot (at least a bit). Follow here: https://github.com/dart-lang/json_serializable/issues/95

**Comment by roughike** [comment]
Incorporated the first feedback round, waiting for @kevmoo to resolve dart-lang/json_serializable#95. Once I can test the simplified way of using `json_serializable`, I'll update the article and ping @kevmoo for a review :)

File: json.md [hunk 16]
```diff
@@ -0,0 +1,415 @@
...
+```dart
+import 'package:build_runner/build_runner.dart';
+import 'build_actions.dart';
+
+main() {
+  watch(buildActions, deleteFilesByDefault: true);
+}
+```
+
+We can run the watcher by doing `dart tool/watch.dart` in our project's root
```
Comment (sethladd) [inline json.md]: s/doing/running

File: json.md [hunk 17]
```diff
@@ -0,0 +1,415 @@
...
+folder. It is safe to start the watcher once and leave it running in the
```
Comment (sethladd) [inline json.md]: please provide an example of how to run this. (it's not obvious how to use the command-line Dart VM that is bundled with Flutter.)

File: json.md [hunk 18]
```diff
@@ -0,0 +1,415 @@
...
+main() async {
+  var result = await build(buildActions, deleteFilesByDefault: true);
+  if (result.status == BuildStatus.failure) {
+    exitCode = 1;
+  }
+}
+```
+
+Now we can generate all the necessary JSON serialization source files by running
+`dart tool/build.dart` in our project's root folder. While the build file is all
```
Comment (sethladd) [inline json.md]: this assumes you have the Dart SDK installed separately from Flutter. Instead, please show how to run the command-line Dart VM that is shipped with the flutter SDK.

File: json.md (hunk 16 above)
Comment (roughike) [reply json.md]: As @kevmoo said [in this comment](https://github.com/flutter/website/pull/773#discussion_r160228505), seems like these sections are going away completely. With the new simplified version of `json_serializable`, there's no need for manually configuring build actions, watchers and build files. Correct me if I'm wrong.

So, should I:
* wait until dart-lang/json_serializable#95 gets resolved and replace these old instructions with new ones, or:
* fix these issues now, publish this article with the old instructions, then update with new `json_serializable` section once dart-lang/json_serializable#95 is resolved?

It's your call @sethladd. I'm fine with either way!

File: json.md (hunk 18 above)
Comment (sethladd) [reply json.md]: Here's how to do it: https://groups.google.com/forum/#!topic/flutter-dev/xoS0WSUax8Q

File: json.md (hunk 16 above)
Comment (sethladd) [reply json.md]: if https://github.com/dart-lang/json_serializable/issues/95 is less than a week away, then let's wait.

If it's more than a week, let's publish this article now, as long as you're OK with updating it "soon".

@kevmoo what's the ETA?

Comment (kevmoo) [reply json.md]: As of Flutter latest, everything seems to work – https://github.com/flutter/flutter/commit/6e38b42919dec025cbeea8ab3e18906e18df23e3

It's all about the very tight dependency constraints on packages in flutter & flutter_test

Comment (roughike) [reply json.md]: Pretty sure we'll want to delay the new documentation until that change is rolled to the alpha branch, which most people are using by default?

(BTW, I have not abandoned this. I would love to do as many feedback rounds as it takes to publish the article, but I'm very busy this week.)

**Comment by Hixie** [comment]
Just for my own purposes (to make it easier to see what's going on when I go through all the open PRs):
* This is waiting for https://github.com/flutter/flutter/pull/14070 to roll into dev
* @roughike is going to make some additional changes as well

BTW, thanks for your contributions @roughike ! Much appreciated.

**Comment by Hixie** [comment]
This should be unblocked now, we just rolled dev.

**Comment by roughike** [comment]
Great! I'm getting back to Finland in 12hours or so, I'll finish the changes once I wake up and recover from the jet lag. I'll do the "what if I decode lots of JSON and the UI lags" part as well.

**Comment by kevmoo** [comment]
Can confirm. With Flutter 0.0.21, you can get all of the latest/greatest
build tools

**Comment by roughike** [comment]
Here's what I found about the "avoiding blocking the main UI isolate when decoding large amounts of JSON" that we discussed in the afterparty:

* [a huge amount of JSON objects](https://gist.githubusercontent.com/roughike/497dbec6708715759f291439610b660e/raw/2281e7cf29b8b22b17ac1cbf9c73a759dc81dcf0/users.json) (really!) are required to produce any noticeable UI lag
* basically, you must choose either having the UI lag for a second or not having any lag but waiting much longer. 

Does the `compute()` function and spawning a new Isolate cause so much overhead that it actually takes more than just deserializing the JSON in the main isolate?

[See the sample project here.](https://github.com/roughike/json_serialization_test)

**Comment by roughike** [comment]
(I'm pretty sure no one would request  15 000 object from a backend API at once, but they would implement paging instead.)

File: json.md [hunk 19]
```diff
@@ -0,0 +1,357 @@
+---
+layout: page
+title: JSON and serialization
+permalink: /json/
+---
+
+It is hard to think of a mobile app that doesn't need to communicate with a web
+server at some point. When making network-connected apps, the chances are that
```
Comment (sethladd) [inline json.md]: server, or easily store structured data, at some point

File: json.md [hunk 20]
```diff
@@ -0,0 +1,357 @@
...
+
+In this tutorial, we look into ways of using JSON with Flutter. We go over what
+JSON solution to use in different scenarios and why.
+
+* TOC Placeholder
+{:toc}
+
+## Which JSON serialization method is right for me?
+
+There isn't a one size fits all JSON solution for every project.
```
Comment (sethladd) [inline json.md]: Add a short paragraph here:

This article covers two general strategies for working with JSON:

* Manual serialization and deserialization
* Automated serialization and deserialization via code generation

File: json.md [hunk 21]
```diff
@@ -0,0 +1,357 @@
...
+
+Also, real-world scenarios are not usually that simple. It is unlikely that we
+can get by with such small JSON responses. Nested JSON objects are not that
+uncommon either.
+
+It would be nice if there were something that handled the JSON serialization for
+us. Luckily, there is!
+
+<a name="code-generation"></a>
+## Serializing JSON using json_serializable
```
Comment (sethladd) [inline json.md]: mention: code generation here (as we talk about it above as "code generation" and that's more what people are looking for. They don't know what json_serializable is , at this point)

For example: serializing JSON using code generation

**Comment by sethladd** [comment]
cc @kevmoo to double-check the json_serializable bits of the article. we're about to commit.

**Comment by roughike** [comment]
Latest changes addressed.

> Lets address the small comments, merge this article. And then lets add a new section called "working with JSON on another thread"

Agreed. I was thinking of including that new section with this PR, but let's ship this as soon as this is ready. This has already been sitting one month, sorry about that :)

I'll create a separate PR about that "working with JSON on another thread".

**Comment by roughike** [comment]
Hey @sethladd and @kevmoo, would it make sense to ship a live template to make using `json_serializable` easier?

Something like this:

json_serializable live template

An even crazier idea would be having a squiggly red line under the class name saying "Code not generated yet". By pressing Alt+Enter, an option saying "Generate code for json_serializable model" would appear. By hitting enter, the following steps would be done:

* Check if the json_serializable library is declared in pubspec file. If not, add it automatically and refresh the dependencies.
* Run `flutter packages pub run build_runner build` for the user to build the generated code.

From the user point of view, all they had to do was write "json", press enter, press Alt-Enter and now they have working `json_serializable` setup in their project.

**Comment by sethladd** [comment]
@roughike that's really cool! I encourage you to open a new issue for that feature request. That way, we can think about how to do it for both VS Code and Android Studio.

**Comment by sethladd** [comment]
Other than that, thank you! LGTM from me.

Pending LGTM from @kevmoo

**Comment by roughike** [comment]
@sethladd What's the correct repo to open that issue? flutter-intellij would be my first guess, but don't know if it's related to Dart analysis server and just disabled on plain Dart projects. :)

File: json.md [hunk 22]
```diff
@@ -0,0 +1,363 @@
...
+
+print('Howdy, ${user.name}!');
+print('We sent the verification link to ${user.email}.');
+```
+
+To serialize a user, we just call the `toJson` method in the `User` class.
+
+<!-- skip -->
+```dart
+String json = JSON.encode(user.toJson());
```
Comment (kevmoo) [inline json.md]: `toJson` is not needed here. `JSON.encode` calls `toJson` for you!

File: json.md [hunk 23]
```diff
@@ -0,0 +1,363 @@
...
+  final String name;
+  final String email;
+
+  User(this.name, this.email);
+
+  User.fromJson(Map<String, dynamic> json)
+      : name = json['name'],
+        email = json['email'];
+
+  Map<String, dynamic> toJson() {
```
Comment (kevmoo) [inline json.md]: Would probably use `=>` notation here.

```dart
Map<String, dynamic> toJson() => 
  {
   'name'
...
```

File: json.md [hunk 24]
```diff
@@ -0,0 +1,363 @@
...
+uncommon either.
+
+It would be nice if there were something that handled the JSON serialization for
+us. Luckily, there is!
+
+<a name="code-generation"></a>
+## Serializing JSON using code generation libraries
+
+Although there are other libraries available, in this tutorial, we use the
+[json_serializable package](https://github.com/dart-lang/json_serializable). It
```
Comment (kevmoo) [inline json.md]: Link to the package site - https://pub.dartlang.org/packages/json_serializable

File: json.md [hunk 25]
```diff
@@ -0,0 +1,363 @@
...
+```dart
+Map userMap = JSON.decode(json);
+var user = new User.fromJson(userMap);
+```
+
+Same goes for serialization. The calling API is the same as before.
+
+<!-- skip -->
+```dart
+String json = JSON.encode(user.toJson());
```
Comment (kevmoo) [inline json.md]: ditto about not needing `.toJson()` here

File: json.md (hunk 23 above)
Comment (roughike) [reply json.md]: I can definitely change it, but the original reasoning for me going the long route was that `dartfmt` always formats it like this:

```dart
Map<String, dynamic> toJson() => {
      'name': name,
      'email': email,
    };
```

Which in my eyes didn't look as pretty. :)

**Comment by kevmoo** [comment]
Not a biggy either way...

**Comment by sethladd** [comment]
@roughike please open the issue here (flutter/flutter/issues) as it is a general request that spans many parts of the system.

Are we ready to merge? Anything else to add , @roughike ?

**Comment by roughike** [comment]
I'm ready to merge. I'll create the issue tomorrow and open a new PR about "serializing JSON on a separate thread" when I can!

**Comment by sethladd** [comment]
Nice, thanks!
